Use multiprocessing to speed up regression tests

goodmami opened the pull request:
Less frequent and detailed reporting of test progress, but ~2x faster.

The time to run all tests for me went from ~4.5 min to ~2.3 min. Can someone else please test this out and report before/after times?

Because the processes are done in parallel, I cannot easily print the detailed `[customizing]`, `[compiling]`, etc. statuses as before, nor what test is currently running, so instead I just print the final result and make the progress bar fill the entire width. I added a counter to the progress bar, as well. It now looks like this:

![Screenshot from 2020-08-02 12-35-33](https://user-images.githubusercontent.com/1428419/89115599-bee76f80-d4bc-11ea-92bf-4f8cb0476bf7.png)

This uses 4 processes (configured with the `PROCESSES` variable), which works for me because I have 4 cores on my processor. I could just not specify the number of processes, which defaults to `os.cpu_count()`, but I don't want to use > 4 (on more powerful systems) because these all share the system memory and I don't want to risk that parses fail because of memory errors. I don't think there's much difference in speed if I request more processes than available processors/cores.

There's also a `BATCH_SIZE` variable which controls how many jobs are processed on each process before reporting, but my experiments did not show any benefit to increasing this. The bottleneck is probably all the disk writing, and there's not much batching can do about that.

goodmami committed:
Use multiprocessing to speed up regression tests

Less frequent and detailed reporting of test progress, but ~2x faster.

dantiston commented:
hah! You beat me to it. I implemented this a couple weeks ago but I didn't get around to fixing the output problem. I think it probably could be improved by keeping track of the running tests and rewriting the status for all of the running tests, and then leaving the finished tests in the terminal, but I think reducing the time is more important than the output.

I personally think we should allow the machine to choose the max CPUs, for instance on a machine with only 2 cores will still result in issues, and then on a machine with 16 cores, it's not as useful. Maybe instead of having the max be a constant, it could be either an environment variable or an CLI argument?

I tried it on my machine, and it worked dandy:
```
real	3m44.191s
user	8m13.135s
sys	2m25.481s
```

goodmami commented:
> hah! You beat me to it.

Oops. Sorry :)

> I think it probably could be improved by keeping track of the running tests and rewriting the status for all of the running tests [...]

Yes, I think we *could* do that, but I'm concerned that the complexity of that solution would outweigh the benefits, and potentially make it less comprehensible to someone down the road. Personally I didn't find the per-step status messages very useful, and once the regression tests start going you don't notice the synchronization delays as much as with the first set or two.

> I personally think we should allow the machine to choose the max CPUs, for instance on a machine with only 2 cores will still result in issues, and then on a machine with 16 cores, it's not as useful. Maybe instead of having the max be a constant, it could be either an environment variable or an CLI argument?

Requesting more processes than CPUs/cores is not a problem; the extra processes just get queued. And like I said, I'm worried that we'd start having OOM errors if we allow it to use all CPUs and don't have enough memory, but I guess these grammars are small and the test sentences are simple (except for Sahaptin, maybe).

I don't want to make it too complex. I'll just make the default `None` which uses `os.cpu_count()`, and hope it doesn't cause issues.

> I tried it on my machine, and it worked dandy:
>
>     real	3m44.191s

So I guess it was ~7m before?

goodmami committed:
Use default number of parallel processes in rtest

goodmami committed:
Add a space after progress bar

When the last column is a non-space it looks like it might have wrapped.

goodmami committed:
Delete ACE .dat file after each regression test

Each of the 478 tests had a .dat file of ~8MB, so this saves nearly
4GB of disk space. The grammar files are not deleted, so it's trivial
to recreate the .dat file using ACE.

goodmami commented:
I bundled another change that's unrelated to multiprocessing but related to overall efficiency. Now the ACE .dat file will be deleted as soon as each test is complete. The grammar files are not deleted, so it can be recreated with `ace -g ace/config.tdl -G grm.dat`. Each grammar file is about ~8MB or so, so this saves about 4GB of disk space.

Another place that could be improved is the log files. Each full run generates nearly 500 new files, so this can grow quickly. In terms of disk space it's not so bad (~5MB for all logs for a single run), but a directory with tens of thousands of files is no fun to work in. There's very limited utility to old logs, so I'd be happy just deleting the old ones before each run (so the user would backup any logs they particularly want to keep), but if others don't like that option we could compress logs from previous days (e.g., `2020-08-03.tgz`, etc.), or do nothing.

olzama commented:
Will it still be possible to use the visual debugger as normal, with this multithreading? (I am not very familiar with using multiple cores.) I am assuming, at any point, if I want to debug a particular test, I should still be able to do it as before, the multithreading won't get in the way?

olzama commented:
> I'd be happy just deleting the old ones before each run (so the user would backup any logs they particularly want to keep), but if others don't like that option we could compress logs from previous days (e.g., `2020-08-03.tgz`, etc.), or do nothing.

Sometimes it is useful to look at older logs, I think (though not often). What I wanted to do at some point was a "--cleanup" option (or something like that) which would delete everything from the logs/ and the grammars/ upon request.

goodmami commented:
> Will it still be possible to use the visual debugger as normal, with this multithreading? (I am not very familiar with using multiple cores.) I am assuming, at any point, if I want to debug a particular test, I should still be able to do it as before, the multithreading won't get in the way?

Sorry I'm not following. What's the visual debugger? Multiprocessing won't prevent you from running a single test, it just means that running multiple tests will take less time because the computer can process several in parallel. The result is the same, and even reported in the same order.

> Sometimes it is useful to look at older logs, I think (though not often).

Right, but in my (limited) experience not as often as the desire to delete the old ones. So my thought is that for those "sometimes" the developer would back up the one they're interested in. But I guess that's not ideal either, because they might only recognize the need to backup a log after it's been deleted :P

> What I wanted to do at some point was a "--cleanup" option (or something like that) which would delete everything from the logs/ and the grammars/ upon request.

Yes, that's probably better.

olzama commented:
Visual debugger - a GUI in the IDE which shows you the state of the program
etc., after the execution stops at a breakpoint.

On Mon, Aug 3, 2020 at 7:21 PM Michael Wayne Goodman <
notifications@github.com> wrote:

> Will it still be possible to use the visual debugger as normal, with this
> multithreading? (I am not very familiar with using multiple cores.) I am
> assuming, at any point, if I want to debug a particular test, I should
> still be able to do it as before, the multithreading won't get in the way?
>
> Sorry I'm not following. What's the visual debugger? Multiprocessing won't
> prevent you from running a single test, it just means that running multiple
> tests will take less time because the computer can process several in
> parallel. The result is the same, and even reported in the same order.
>
> Sometimes it is useful to look at older logs, I think (though not often).
>
> Right, but in my (limited) experience not as often as the desire to delete
> the old ones. So my thought is that for those "sometimes" the developer
> would back up the one they're interested in. But I guess that's not ideal
> either, because they might only recognize the need to backup a log after
> it's been deleted :P
>
> What I wanted to do at some point was a "--cleanup" option (or something
> like that) which would delete everything from the logs/ and the grammars/
> upon request.
>
> Yes, that's probably better.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/delph-in/matrix/pull/493#issuecomment-668341048>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACTURKUAI2VK3PGV3MGGX2DR65WAFANCNFSM4PSKUTVA>
> .
>
-- 
Olga Zamaraeva

olzama commented:
So, this kind of thing:

<img width="1237" alt="Screen Shot 2020-08-03 at 7 37 06 PM" src="https://user-images.githubusercontent.com/10963114/89246825-d14ad180-d5c0-11ea-89e3-9ff5988ea33c.png">

goodmami commented:
I see. I never really use those. Can you test it out and see what happens? A quick search suggests that some debuggers are better than others at traversing forked processes.

olzama commented:
I will look... (Personally, I simply won't be able to work without one! To me being able to debug like this is being able to program at all, almost...)

dantiston commented:
@olzama is that PyCharm? It looks like PyCharm has supported multiprocessing debugging for ~7 years, so you should be good 🙂 (https://youtrack.jetbrains.com/issue/PY-8258)

olzama commented:
I think** that what's important here is that it's just separate multiple processes and not "threads". So long as there are no threads, I believe** the debugger should not get too confused (and the situation overall would not get too complex, with only modest speed gain).

**My wonderful teacher Martin Tompa opened his lecture on concurrency as follows. He said: "The department forced me to include this topic in this course. We will have two lectures and an assignment. But what I want you to actually remember from this is the following: when you go to an interview and they ask you what you know about concurrency and parallelism, I want you to tell them: "Absolutely nothing".

goodmami commented:
Ok it sounds like we're good? I'll merge it then.

olzama commented:
Well, I haven't checked yet :). I hope it will be OK :). I appreciate the speed up, and yes, the tests were getting a bit slow... so something had to be done at some point... I am just worried a bit that perhaps a modest gain in speed is not worth introducing concurrency, as it is kind of notorious for introducing unexpected problems which go unnoticed for years...

goodmami commented:
Ok, if you have trouble let's maybe add a --debug option that disables multiprocessing. Should be easy:

```python
if args.debug:
    for test in tests:
        name, result, logpath = run_test(test)
        ...
else:
    with multiprocess.Pool(PROCESSES) as pool:
        ...
```

goodmami commented:
Ok the situation isn't as rosy as I thought. I tried with Visual Studio Code, which also has a debugger, and it doesn't seem to work with multiprocessing, or at least how we're calling it. I'll see about making a `--debug` option in the other PR.

olzama commented:
Could we please have a tag for the last version which doesn't use multiple processes? :) It seems an important enough transition to me. (Btw I deleted the tag I added for information structure tests.)

goodmami commented:
I mean, I wouldn't but I can't stop you. And we can always delete it later. But do you plan on switching back to that tag each time you need to debug something? Why don't we just fix whatever the problem is or find an easier workaround.

olzama commented:
I just want to have a simple way of finding the last version which does not use multiprocessing at all in case multiprocessing causes unexpected problems some (possibly very distant) time in the future (which I think concurrency is notorious for, causing unexpected problems which go unnoticed for years). I trust you but I still don't trust multiprocessing :). Consider also that, some years from now, some beginner student may need to deal with this.

olzama commented:
> And we can always delete it later.

That would not be my intention though; to me introducing concurrency is a big deal :).
